### PR TITLE
Introduce UnpackStrategyType

### PR DESCRIPTION
**Pull request by dduugg**
- [ ] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [ ] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [ ] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [ ] Have you successfully run `brew style` with your changes locally?
- [ ] Have you successfully run `brew typecheck` with your changes locally?
- [ ] Have you successfully run `brew tests` with your changes locally?

-----

### CODE REVIEW

File: Library/Homebrew/unpack_strategy.rb [hunk 1]
```diff
@@ -6,46 +6,27 @@
 # Module containing all available strategies for unpacking archives.
 module UnpackStrategy
   extend T::Helpers
+  include SystemCommand::Mixin
+  abstract!
+
+  # rubocop:disable Style/MutableConstant
```
Comment (MikeMcQuaid) [inline Library/Homebrew/unpack_strategy.rb:12]:
```suggestion
  # A comment explaining why
  # rubocop:disable Style/MutableConstant
```

File: Library/Homebrew/unpack_strategy.rb [hunk 2]
```diff
@@ -6,46 +6,27 @@
...
+  UnpackStrategyImpl = T.type_alias { T.all(T::Class[UnpackStrategy], UnpackStrategy::ClassMethods) }
```
Comment (MikeMcQuaid) [inline Library/Homebrew/unpack_strategy.rb:13]:
```suggestion
  UnpackStrategyType = T.type_alias { T.all(T::Class[UnpackStrategy], UnpackStrategy::ClassMethods) }
```

or something? Not a huge fan of `Impl` here.

**Review by MikeMcQuaid — COMMENTED** [review]
Great work so far here, thanks @dduugg!

**Review by issyl0 — APPROVED** [review]
I think I understand this more now I've read more things. :shipit:

**Comment by issyl0** [comment]
Since this is merging into #17636 I'll rebase-merge this in now and we can deal with the (pre-existing) failing `cask` tests over there.

**Pull request merged**